### DISCUSSION TITLE
Key value error when iterating through edges

### DISCUSSION THREAD

**Discussion by Fedebindi97** [networkx/networkx#4771] [General]
Hello everybody. I'm currently developing an agent-based model that features a network in which agents dynamically create links between each other. These links have some attributes, among which "expiry". 

At a certain point in the code, I need to create lists of all in_edges and out_edges pertaining to a certain node ("self", as I'm creating a class method; "model" is another class which is passed into the agent as an __init__ argument, and which contains the network) for which "expiry" takes certain values. I use list comprehension:

`
out_contracts_2_3=[(a,b) for a,b in self.model.modelgraph.out_edges(self) if self.model.modelgraph[a][b]["expiry"]==2 or self.model.modelgraph[a][b]["expiry"]==3]
`

`
in_contracts_2_3=[(a,b) for a,b in self.model.modelgraph.in_edges(self) if self.model.modelgraph[a][b]["expiry"]==2 or self.model.modelgraph[a][b]["expiry"]==3]
`

At the first turn, the list is empty: the model is initialized with no links among the nodes. When I run the code, Python immediately raises 

`KeyError: 'expiry'`

At first, I thought that the problem was the emptiness of the list, so I tried to enclose the above code in an if statement, that would not consider the above line of codes if `[(a,b) for a,b in self.model.modelgraph.out_edges(self)]` or `[(a,b) for a,b in self.model.modelgraph.in_edges(self)]` were empty lists, but the error is raised anyway. The funny thing is that the error is raised only for the second line of code, despite the two being exactly the same.

Does anyway have any idea of what's going on? It may be a naive issue, but I'm really a Python beginner...

Thanks!

**Comment by MridulS** [comment]
The  `KeyError` (`KeyError: 'expiry'`)  pops up when there is no key in the attribute dictionary, i.e the attribute `expiry` is not present in the edge of the `modelgraph`. I think (just a speculation) this is happening only in one case because the key `expiry` is not set properly for all the edges in the graph. A quick example:


``` python

In [1]: import networkx as nx

In [2]: G = nx.DiGraph()

In [3]: G.add_edge(1, 2, expiry=1) # the outgoing edge from node 1 has the attribute expiry set to 1

In [4]:  G.add_edge(2, 1) # the attribute expiry is missing in the incoming edge.

In [5]: [(a, b) for a, b in G.out_edges(1) if G[a][b]['expiry'] == 1]
Out[5]: [(1, 2)]

In [6]: [(a, b) for a, b in G.in_edges(1) if G[a][b]['expiry'] == 1]
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-6-45b0424c3541> in <module>
----> 1 [(a, b) for a, b in G.in_edges(1) if G[a][b]['expiry'] == 1]

<ipython-input-6-45b0424c3541> in <listcomp>(.0)
----> 1 [(a, b) for a, b in G.in_edges(1) if G[a][b]['expiry'] == 1]

KeyError: 'expiry'
```

**Reply by Fedebindi97** [reply]
But the point is that the line of code that creates the links correctly sets the expiry attribute:

`self.model.modelgraph.add_edge(self, other_agent, weight=contract.protection/30, expiry=contract.expiry, spread=contract.spread, active=True)`

`contract` is a user-defined object, and `other_agent` is another agent in the model. I can see that in the first simulation step some links are created, but then when the model selects an agent that already has an in_link, it fails to retrieve its `expiry`

**Reply by dschult** [reply]
OK... sounds like @MridulS has identified the problem and you need to track down (inside the rest of your code) why you have an edge that doesn't have the 'expiry' attribute set.

**Reply by Fedebindi97** [reply]
I've made some further checks, and I'm even more puzzled than before. After a single step of the model (which raised the KeyError above), I noticed that there were actually some links (which is weird enough, because in order to form a link agents need to parse edges according to their expiry as described above). I tried to type:

`for a,b,c in model.modelgraph.edges.data("expiry"):`

   ` print(c)`

And since there were two links, I correctly got a couple of 3s (all links are born with expiry=3 in the model). So, links do have an expiry. Furthermore, in a very similar model I created a while ago, the first code I used works perfectly fine.

**Reply by MridulS** [reply]
Hi @Fedebindi97 is the code publicly available? It looks like there are lot of moving parts and debugging this would be hard over GitHub discussions. Or if possible please post a minimal example of the errors :)

**Reply by Fedebindi97** [reply]
No, it's not public, and I don't know how to publish it to Github... I can post it, but it uses another package, mesa (agent-based modeling package for Python). Are you familiar with it?

**Reply by dschult** [reply]
I've used Mesa, but that wouldn't add edges except where your code tells it to. Are you using NetworkX oriented special classes within Mesa that someone else wrote?

If not, try putting print statements into your code to keep track of the edge attributes. For example in the line just before the one that raises the error, use something like:
```python
bad_edges = [u,v,d for u,v,d in model.model_graph.edges.data() if 'expiry' not in d]
print("edges without expiry: ", bad_edges)
```
From that you should learn which edges are causing the exception. Run your code with a set "seed" for the random number generator so you get the same edges every time.

You might also try searching all your code for the string "add_edge" to find where edges are added.

**Reply by Fedebindi97** [reply]
I tried your solution @dschult , but it just prints empty lists. So there is `expiry`, but somehow networkx fails to retrieve it with the method I was trying before. It seems that tuple unpacking works, so I'll just go with that (I don't want to waste your time!) but it really puzzles me that networkx isn't working in this particular circumstance. Anyway no, the code is mine, and there's just one line that create links, so it really felt under control. I'm basically endowing the model with a network, and then each agent (in the step() method) can match with any other agent and form a link that will be added to the model, according to certain rules.
